### PR TITLE
Feature/tech 77  lfg

### PR DESCRIPTION
**Pull request by cybtachyon**

### CODE REVIEW

File: guildOptions.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package main
+
+type GuildOptions struct {
```
Comment (Darcness) [inline guildOptions.go]: There's already a structure for this in /config/config.go

File: lfg.go [hunk 2]
```diff
@@ -0,0 +1,2 @@
+package main
```
Comment (Darcness) [inline lfg.go]: Is this for the lfg command?  There's a handful of folders with the commands already in them.

File: main.go [hunk 3]
```diff
@@ -3,14 +3,22 @@ package main
...
-
 	"github.com/bwmarrin/discordgo"
+	"github.com/ghodss/yaml"
+)
+
+var (
+	commandPrefix string
+	botID         string
+	botname       string
+	gameLib       map[string]string
```
Comment (Darcness) [inline main.go]: Looks like a bad merge.  Most of these are no longer necessary.

File: main.go [hunk 4]
```diff
@@ -24,8 +32,25 @@ func main() {
 	utils.Assert("error creating discord session", err)
 
 	discord.AddHandler(coreMessageHandler)
+
+	user, err := discord.User("@me")              //grabbing account information
+	utils.Assert("error retrieving account", err) //check if error occurred
+	botID = user.ID                               //botID is a variable set to the bots information
```
Comment (Darcness) [inline main.go]: I handle these on line 63

File: main.go [hunk 5]
```diff
@@ -43,11 +68,26 @@ func main() {
 
 		// Where are we?
 		servers := discord.State.Guilds
-		fmt.Printf("Servers (%d):\n", len(servers))
-		for _, server := range servers {
-			fmt.Printf("%s - %s\n", server.Name, server.ID)
+		fmt.Printf("LANFest Bot has started on %d servers.\n", len(servers))
+		fmt.Printf("\nServer Details\n")
+		for i := 0; i < len(discord.State.Guilds); i++ {
```
Comment (Darcness) [inline main.go]: Why ditch the foreach already here?

File: main.go [hunk 6]
```diff
@@ -57,8 +97,45 @@ func main() {
...
+			println(currentrole.ID)
+			if currentrole.Name == "admin" {
+				println(messagechannel)
+			}
+		}
+		ownerchannel, err := discord.UserChannelCreate(currentguild.OwnerID)
+		utils.Assert("error creating channel", err)
+		message, err := discord.ChannelMessageSend(ownerchannel.ID, "Welcome to the LANFest Discord bot! You are registered as the owner of "+currentguild.Name+", so you will need to answer a few questions to complete setup. If you’d rather not, please type the discord ID of another admin-level access user (e.g. "+botname+botID+") to complete setup for you. Type y to continue.")
+		utils.Assert("error sending message", err)
+		println(message.ID)
```
Comment (Darcness) [inline main.go]: You're already doing this in the 'onReady' handler above (starting with line 52).  Still necessary to do it here as well?

File: main.go [hunk 7]
```diff
@@ -57,8 +97,45 @@ func main() {
...
+	}
+
 	defer discord.Close()
 
+	commandPrefix = "!"
```
Comment (Darcness) [inline main.go]: No longer necessary, this is handled in /data/data.go:Constants()

File: main.go [hunk 8]
```diff
@@ -24,8 +32,25 @@ func main() {
...
+	}
+	if err != nil || yamlErr != nil {
+		fmt.Printf("%s %s Error loading game library config file. LFG and game matching features will not work.\n", err, yamlErr)
+	}
+
+	discord.AddHandler(commandHandler) // a listener that when it picks up a message create it runs the function
+
 	discord.AddHandler(func(discord *discordgo.Session, ready *discordgo.Ready) {
-		err = discord.UpdateStatus(0, "Insaniquarium!")
+		err = discord.UpdateStatus(0, "with your heart <3")
```
Comment (Darcness) [inline main.go]: At this point, this code is beefy enough to be pulled into its own handler function and moved below (similar to how commandHandler is structured)

File: main.go [hunk 9]
```diff
@@ -57,8 +97,45 @@ func main() {
...
+				utils.Assert("error sending message", err)
+				println(message.ID)
+			}
+		}
+		for r := 0; r < len(currentguild.Roles); r++ {
+			currentrole := currentguild.Roles[r]
+			println(currentrole.Name)
+			println(currentrole.ID)
+			if currentrole.Name == "admin" {
+				println(messagechannel)
```
Comment (Darcness) [inline main.go]: At this point in execution, 'messagechannel' is merely the last channelID in the list of channels from the server.  Is that the goal here?

File: main.go [hunk 10]
```diff
@@ -24,8 +32,25 @@ func main() {
...
+	gameLibFile, err := ioutil.ReadFile("games.yml")
+	var yamlErr error
+	if err != nil {
+		yamlErr = yaml.Unmarshal(gameLibFile, &gameLib)
+	}
+	if err != nil || yamlErr != nil {
+		fmt.Printf("%s %s Error loading game library config file. LFG and game matching features will not work.\n", err, yamlErr)
+	}
+
+	discord.AddHandler(commandHandler) // a listener that when it picks up a message create it runs the function
```
Comment (Darcness) [inline main.go]: Bad merge again.  The main message handler is named 'coreMessageHandler', and we add it as a handler on line 34.

File: main.go [hunk 11]
```diff
@@ -72,10 +149,27 @@ func coreMessageHandler(session *discordgo.Session, message *discordgo.MessageCr
 		// It's not a command, nothing to do here.
 		return
 	}
+}
+
+func commandHandler(discord *discordgo.Session, message *discordgo.MessageCreate) {
+	user := message.Author
+	if user.ID == botID || user.Bot {
+		return
+	}
```
Comment (Darcness) [inline main.go]: coreMessageHandler already handles this.  Is there a reason to separate them?

File: main.go [hunk 12]
```diff
@@ -72,10 +149,27 @@ func coreMessageHandler(session *discordgo.Session, message *discordgo.MessageCr
...
 	fmt.Printf("Message: %+v || From: %s\n", message.Message, message.Author)
+
+	if strings.HasPrefix(message.Message.Content, "!lfg ") {
+		fmt.Printf("LFG Command string found.\n")
+		specialCharsRegex, err := regexp.Compile("[^a-zA-Z0-9]+")
+		if err != nil {
+			discord.MessageReactionAdd(message.Message.ChannelID, message.Message.ID, "632789111958274068")
+		}
+		gameSuggestion := strings.ToLower(specialCharsRegex.ReplaceAllString(strings.TrimPrefix(message.Message.Content, "!lfg "), ""))
+		discord.ChannelMessageSend(message.Message.ChannelID, fmt.Sprintf("%s has been added to the LFG queue for %s for the next 4 hours.", message.Message.Author.Username, gameSuggestion))
```
Comment (Darcness) [inline main.go]: This should be built into a separate command function (see /commands/chapter.go:PartyOnCommandHandler() for an example).

File: main.go (hunk 8 above)
Comment (Darcness) [reply main.go]: My other PR takes care of splitting this out.

**Comment by cybtachyon** [comment]
Needs to be rebased on master. 👍

**Pull request merged**